// @aws-amplify/graphql-transformer-core/lib/cdk-compat/schema-asset.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransformerSchema = void 0;
const aws_appsync_1 = require("@aws-cdk/aws-appsync");
const core_1 = require("@aws-cdk/core");
const file_asset_1 = require("./file-asset");
class TransformerSchema {
    constructor() {
        this.definition = '';
        this.bind = (api) => {
            if (!this.schemaConstruct) {
                const schema = this;
                this.api = api;
                this.schemaConstruct = new aws_appsync_1.CfnGraphQLSchema(api, 'TransformerSchema', {
                    apiId: api.apiId,
                    definitionS3Location: core_1.Lazy.string({
                        produce: () => {
                            const asset = schema.addAsset();
                            return asset.s3Url;
                        },
                    }),
                });
            }
            return this.schemaConstruct;
        };
        this.addAsset = () => {
            if (!this.api) {
                throw new Error('Schema not bound');
            }
            if (!this.asset) {
                this.asset = new file_asset_1.FileAsset(this.api, 'schema', { fileName: 'schema.graphql', fileContent: this.definition });
            }
            return this.asset;
        };
        this.addToSchema = (addition, delimiter) => {
            const sep = delimiter !== null && delimiter !== void 0 ? delimiter : '';
            this.definition = `${this.definition}${sep}${addition}\n`;
        };
    }
}
exports.TransformerSchema = TransformerSchema;
