// @aws-amplify/graphql-transformer-core/lib/utils/defaultSchema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_SCHEMA_DEFINITION = void 0;
const graphql_1 = require("graphql");
exports.DEFAULT_SCHEMA_DEFINITION = {
    kind: graphql_1.Kind.SCHEMA_DEFINITION,
    directives: [],
    operationTypes: [
        {
            kind: graphql_1.Kind.OPERATION_TYPE_DEFINITION,
            operation: 'query',
            type: {
                kind: graphql_1.Kind.NAMED_TYPE,
                name: {
                    kind: graphql_1.Kind.NAME,
                    value: 'Query',
                },
            },
        },
        {
            kind: graphql_1.Kind.OPERATION_TYPE_DEFINITION,
            operation: 'mutation',
            type: {
                kind: graphql_1.Kind.NAMED_TYPE,
                name: {
                    kind: graphql_1.Kind.NAME,
                    value: 'Mutation',
                },
            },
        },
        {
            kind: graphql_1.Kind.OPERATION_TYPE_DEFINITION,
            operation: 'subscription',
            type: {
                kind: graphql_1.Kind.NAMED_TYPE,
                name: {
                    kind: graphql_1.Kind.NAME,
                    value: 'Subscription',
                },
            },
        },
    ],
};
