// amplify-cli-core/lib/errors/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResourceCountLimitExceedError = exports.ResourceDoesNotExistError = exports.DiagnoseReportUploadError = exports.DebugConfigValueNotSetError = exports.ExportedStackNotInValidStateError = exports.ExportedStackNotFoundError = exports.ExportPathValidationError = exports.CustomPoliciesFormatError = exports.ConfigurationError = exports.UnrecognizedFrontendError = exports.UnrecognizedFrameworkError = exports.AngularConfigNotFoundError = exports.InvalidSubCommandError = exports.MissingParametersError = exports.UnknownArgumentError = exports.UnknownResourceTypeError = exports.ResourceCredentialsNotFoundError = exports.ResourceAlreadyExistsError = exports.NotImplementedError = void 0;
class NotImplementedError extends Error {
}
exports.NotImplementedError = NotImplementedError;
class ResourceAlreadyExistsError extends Error {
}
exports.ResourceAlreadyExistsError = ResourceAlreadyExistsError;
class ResourceCredentialsNotFoundError extends Error {
}
exports.ResourceCredentialsNotFoundError = ResourceCredentialsNotFoundError;
class UnknownResourceTypeError extends Error {
}
exports.UnknownResourceTypeError = UnknownResourceTypeError;
class UnknownArgumentError extends Error {
}
exports.UnknownArgumentError = UnknownArgumentError;
class MissingParametersError extends Error {
}
exports.MissingParametersError = MissingParametersError;
class InvalidSubCommandError extends Error {
}
exports.InvalidSubCommandError = InvalidSubCommandError;
class AngularConfigNotFoundError extends Error {
}
exports.AngularConfigNotFoundError = AngularConfigNotFoundError;
class UnrecognizedFrameworkError extends Error {
}
exports.UnrecognizedFrameworkError = UnrecognizedFrameworkError;
class UnrecognizedFrontendError extends Error {
}
exports.UnrecognizedFrontendError = UnrecognizedFrontendError;
class ConfigurationError extends Error {
}
exports.ConfigurationError = ConfigurationError;
class CustomPoliciesFormatError extends Error {
}
exports.CustomPoliciesFormatError = CustomPoliciesFormatError;
class ExportPathValidationError extends Error {
}
exports.ExportPathValidationError = ExportPathValidationError;
class ExportedStackNotFoundError extends Error {
}
exports.ExportedStackNotFoundError = ExportedStackNotFoundError;
class ExportedStackNotInValidStateError extends Error {
}
exports.ExportedStackNotInValidStateError = ExportedStackNotInValidStateError;
class DebugConfigValueNotSetError extends Error {
}
exports.DebugConfigValueNotSetError = DebugConfigValueNotSetError;
class DiagnoseReportUploadError extends Error {
}
exports.DiagnoseReportUploadError = DiagnoseReportUploadError;
class ResourceDoesNotExistError extends Error {
    constructor(errMessage) {
        super();
        this.name = 'ResourceDoesNotExistError';
        this.message = errMessage;
        this.stack = undefined;
    }
}
exports.ResourceDoesNotExistError = ResourceDoesNotExistError;
class ResourceCountLimitExceedError extends Error {
    constructor(errMessage) {
        super();
        this.name = 'ResourceCountLimitExceedError';
        this.message = errMessage;
        this.stack = undefined;
    }
}
exports.ResourceCountLimitExceedError = ResourceCountLimitExceedError;
